src/Features/Core/Portable/Snippets/AbstractConsoleSnippetProvider.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    internal abstract class AbstractConsoleSnippetProvider : AbstractSnippetProvider
+    {
+        protected abstract bool ShouldDisplaySnippet(SyntaxContext context);
+        protected abstract SyntaxNode? GetAsyncSupportingDeclaration(SyntaxToken token);
+
+        public AbstractConsoleSnippetProvider()
+        {
+        }
+
+        public override string SnippetIdentifier => "cw";
+
+        public override string SnippetDisplayName => FeaturesResources.Write_to_the_Console;
+
+        protected override async Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var symbol = await GetSymbolFromMetaDataNameAsync(document, cancellationToken).ConfigureAwait(false);
```

Comment:
consider renaming method/local for clarity.  i presume this is the Console symbol? if so, consider putting in the name.